PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/AddAzureDependencyAction.java

```diff
@@ -0,0 +1,210 @@
+        assert mavenProject != null;
+
+        MavenEmbeddersManager embeddersManager = projectsManager.getEmbeddersManager();
+        MavenExplicitProfiles profiles = mavenProject.getActivatedProfilesIds();
+        MavenProjectsProcessor myFoldersResolvingProcessor = new MavenProjectsProcessor(project, "Generating Maven sources...", true, embeddersManager);
+        MavenEmbedderWrapper embedder = embeddersManager.getEmbedder(mavenProject, MavenEmbeddersManager.FOR_DEPENDENCIES_RESOLVE);
+
+        ProgressManager.getInstance().runProcessWithProgressSynchronously(() -> {
+            ProgressIndicator progressIndicator = ProgressManager.getInstance().getProgressIndicator();
+            progressIndicator.setText("Check existing dependencies");
+            myFoldersResolvingProcessor.scheduleTask((thisProject, mavenEmbeddersManager, mavenConsole, mavenProgressIndicator) -> {
+                embedder.customizeForResolve(mavenConsole, mavenProgressIndicator);
+                embedder.clearCachesFor(mavenProject.getMavenId());
+                final String evaluateEffectivePom = embedder.evaluateEffectivePom(mavenProject.getFile(),
+                        profiles.getEnabledProfiles(), profiles.getDisabledProfiles());
+                if (StringUtils.isEmpty(evaluateEffectivePom)) {
+                    PluginUtil.showErrorNotification("Error", "Failed to evaluate effective pom.");
+                    return;
+                }
+                try {
+                    SpringCloudDependencyManager manager = new SpringCloudDependencyManager(evaluateEffectivePom);
+
+                    Map<String, String> versionMap = manager.getCurrentVersions(GROUP_ID);
+                    Map<String, String> springBootVersionMap = manager.getCurrentVersions(SPRING_BOOT_GROUP_ID);
+                    Map<String, String> springCloudVersionMap = manager.getCurrentVersions(SPRING_CLOUD_GROUP_ID);
+                    Map<String, Map<String, String>> versionMaps = new HashMap<>();
+                    versionMaps.put(GROUP_ID, versionMap);
+                    versionMaps.put(SPRING_BOOT_GROUP_ID, springBootVersionMap);
+                    versionMaps.put(SPRING_CLOUD_GROUP_ID, springCloudVersionMap);
+                    ProgressManager.checkCanceled();
+                    String springBootVer = springBootVersionMap.get("spring-boot-autoconfigure");
+                    if (StringUtils.isEmpty(springBootVer)) {
+                        throw new AzureExecutionException(String.format("Module %s is not a spring-boot application.", module.getName()));
+                    }
+                    ProgressManager.checkCanceled();
+
+                    progressIndicator.setText("Get latest versions ...");
+                    List<DependencyArtifact> dep = new ArrayList<>();
+                    dep.add(getDependencyArtifact(GROUP_ID, ARTIFACT_ID, versionMaps));
+                    ProgressManager.checkCanceled();
```

Comment:
do we need to call `ProgressManager.checkCanceled();` after add one dependencyArtifact? If `getDependencyArtifact` cost to much time, shall we move the `ProgressManager.checkCanceled();` to the method?

Reply:
Resolved getDependencyArtifact will not cost too much time